openedx/core/djangoapps/bookmarks/models.py

```diff
@@ -239,7 +240,15 @@ def create(cls, data):
 
         data['course_key'] = usage_key.course_key
 
-        xblock_cache, created = cls.objects.get_or_create(usage_key=usage_key, defaults=data)
+        try:
+            # get_or_create is vulnerable to a race condition which can cause
+            # IntegrityErrors to be raised here.
+            # See: https://code.djangoproject.com/ticket/13906 and https://code.djangoproject.com/ticket/18557.
+            xblock_cache, created = cls.objects.get_or_create(usage_key=usage_key, defaults=data)
+        except IntegrityError:
+            # get_or_create should not fail twice. The object responsible for the
+            # previously raised IntegrityError should be retrieved.
+            xblock_cache, created = cls.objects.get_or_create(usage_key=usage_key, defaults=data)
```

Comment:
@muzaffaryousaf This strategy does not work inside views wrapped in transactions. The second `get` will still result in the same error.

I better solution would be prevent `IntegrityErrors` by disabling the button while an existing request is in flight.